src/include/Visitor.h

```diff
@@ -1,10 +1,34 @@
 #pragma once
 
-#include <Parser.h>
-#include <ASTree.h>
+namespace AST {
+    // hacky way around C++ limitations
```

Comment:
what exactly is the limitation?

Reply:
So because C/C++ compilers were traditionally single pass, you need to declare things before they’re used (e.g. Visitor needs to know about the ASTree types). Like, literally above them, in the code. Normally you would use #include with a header file at the top to get those declarations. But ASTree also needs to know about Visitor. So the two files would need to include each other. But then they recursively include each other, in that ASTree has Visitor which has ASTree which has... This would cause issues with multiple definition etc, which #pragma once puts a stop to. But #pragma once works by literally throwing out an #include that’s already been seen. So we’re back to square one, where at least one of the two classes hasn’t seen the declaration of the other. So the way around this is to just forward declare the ASTree types in Visitor, and leave it up to the linker to figure out where those types are defined.

Reply:
tl;dr the ASTree types need to be declared before Visitor is defined and vice versa. The easiest way around this is to just use a forward declaration.

Reply:
Oh, and why don’t we forward declare Visitor in ASTree rather than the other way around? Because ASTree actually depends on the implementation (it needs to call the visit function defined by Visitor) the compiler will complain about “access to incomplete type”. But that’s not a problem for Visitor which, being completely abstract, needs to know the ASTree types exist but doesn’t actually do anything to them in any way.

Reply:
Ah I see, that's pretty interesting!